samples/net/cellular_modem/src/main.c

```diff
@@ -0,0 +1,245 @@
+	int ret;
+
+	sample_test_dns_in_progress = true;
+
+	ret = dns_get_addr_info(SAMPLE_TEST_ENDPOINT_HOSTNAME,
+				DNS_QUERY_TYPE_A,
+				&dns_id,
+				sample_dns_request_result,
+				NULL,
+				19000);
+	if (ret < 0) {
+		return -EAGAIN;
+	}
+
+	if (k_sem_take(&dns_query_sem, K_SECONDS(20)) < 0) {
+		return -EAGAIN;
+	}
+
+	return 0;
+}
+
+int sample_echo_packet(struct sockaddr *ai_addr, socklen_t ai_addrlen)
+{
+	int ret;
+	int socket_fd;
+
+	printk("Opening UDP socket\n");
+
+	socket_fd = zsock_socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+
+	if (socket_fd < 0) {
+		printk("Failed to open socket\n");
+		return -1;
+	}
+
+	printk("Socket opened\n");
+
+	if (ai_addr->sa_family == AF_INET) {
+		net_sin(ai_addr)->sin_port = htons(SAMPLE_TEST_ENDPOINT_UDP_ECHO_PORT);
+	} else if (ai_addr->sa_family == AF_INET6) {
+		net_sin6(ai_addr)->sin6_port = htons(SAMPLE_TEST_ENDPOINT_UDP_ECHO_PORT);
+	} else {
+		printk("Unsupported address family\n");
+		return -1;
+	}
+
+	printk("Sending sample test packet\n");
+
+	ret = zsock_sendto(socket_fd, sample_test_packet, sizeof(sample_test_packet), 0,
+			   ai_addr, ai_addrlen);
+
+	if (ret < sizeof(sample_test_packet)) {
+		printk("Failed to send sample test packet\n");
+		return -1;
+	}
+
+	printk("Await sample test packet echo\n");
+
+	ret = zsock_recv(socket_fd, sample_recv_buffer, sizeof(sample_recv_buffer), 0);
+
+	if (ret != sizeof(sample_test_packet)) {
+		printk("Echoed sample test packet has incorrect size\n");
+		return -1;
+	}
+
+	if (memcmp(sample_test_packet, sample_recv_buffer, sizeof(sample_recv_buffer)) != 0) {
+		printk("Echoed sample test packet data mismatch\n");
+		return -1;
+	}
+
+	printk("Echo recevied successfully!\n");
+
+	printk("Close UDP socket\n");
+
+	ret = zsock_close(socket_fd);
+
+	if (ret < 0) {
+		printk("Failed to close socket\n");
+		return -1;
+	}
+
+	return 0;
+}
+
+int main(void)
+{
+	uint32_t raised_event;
+	const void *info;
+	size_t info_len;
+	int ret;
+
+	init_sample_test_packet();
+
+	printk("Powering on modem\n");
+
+	pm_device_action_run(modem, PM_DEVICE_ACTION_RESUME);
+
+	printk("Bring up network interface\n");
+
+	ret = net_if_up(net_if_get_default());
+
+	if (ret < 0) {
+		printk("Failed to bring up network interface\n");
+		return -1;
+	}
+
+	printk("Waiting for L4 connected\n");
+
+	ret = net_mgmt_event_wait_on_iface(net_if_get_default(),
+					   NET_EVENT_L4_CONNECTED, &raised_event, &info,
+					   &info_len, K_SECONDS(120));
+
+	if (ret != 0) {
+		printk("L4 was not connected in time\n");
+		return -1;
+	}
+
+	printk("Performing DNS lookup of %s\n", SAMPLE_TEST_ENDPOINT_HOSTNAME);
+
+	ret = sample_dns_request();
+
+	if (ret < 0) {
+		printk("DNS query failed\n");
+		return -1;
+	}
+
+	ret = sample_echo_packet(&sample_test_dns_addrinfo.ai_addr,
+				 sample_test_dns_addrinfo.ai_addrlen);
+
+	if (ret < 0) {
+		printk("Failed to send echo\n");
+		return -1;
+	}
+
+	printk("Restart modem\n");
+
+	ret = pm_device_action_run(modem, PM_DEVICE_ACTION_SUSPEND);
```

Comment:
I see this cause a deadlock in the system waiting on  a semaphore forever and everything dies.

Reply:
The driver is quite naive ATM, and will deadlock if it gets stuck in a state where it does not respond to the power down event. The plan is to improve error handling over time.

Reply:
Makes sense, I think we should fix the power off state to also account for the reset line. If there is a reset line for the modem, just hold it in reset.